pulp_container/app/tasks/builder.py

```diff
@@ -76,20 +76,26 @@ def add_image_from_directory_to_repository(path, repository, tag):
     with repository.new_version() as new_repo_version:
         new_repo_version.add_content(Manifest.objects.filter(pk=manifest.pk))
         new_repo_version.add_content(Tag.objects.filter(pk=tag.pk))
-        with open(manifest_artifact.file.path, "r") as manifest_file:
-            manifest_json = json.load(manifest_file)
-            config_blob = get_or_create_blob(manifest_json["config"], manifest, path)
-            manifest.config_blob = config_blob
-            manifest.save()
-            new_repo_version.add_content(Blob.objects.filter(pk=config_blob.pk))
-            for layer in manifest_json["layers"]:
-                blob = get_or_create_blob(layer, manifest, path)
-                new_repo_version.add_content(Blob.objects.filter(pk=blob.pk))
+
+        manifest_json = json.load(manifest_artifact.file)
+        manifest_artifact.file.close()
+
+        config_blob = get_or_create_blob(manifest_json["config"], manifest, path)
+        manifest.config_blob = config_blob
+        manifest.save()
+        new_repo_version.add_content(Blob.objects.filter(pk=config_blob.pk))
+
+        blobs_to_add = []
+        for layer in manifest_json["layers"]:
+            blobs_to_add.append(get_or_create_blob(layer, manifest, path))
+
+        new_repo_version.add_content(Blob.objects.filter(pk__in=blobs_to_add))
```

Comment:
you are still calling add_content here 4 times. collect all the content pks are i've suggested in previous comment